File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.java
import java.io.*;
import java.util.Arrays;
import java.util.PriorityQueue;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        int num = Integer.parseInt(br.readLine());
        StringBuilder ans = new StringBuilder();
        ans.append(num).append("\n");

        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>();
        if (n >= 2) {
            int num2 = Integer.parseInt(br.readLine());
            int mn = Math.min(num, num2);
            int mx = Math.max(num, num2);

            minHeap.add(-mn);
            maxHeap.add(mx);

            ans.append(mn).append("\n");
        }

        for (int unused = 2; unused < n; unused++) {
            num = Integer.parseInt(br.readLine());

            int[] three = {num, -minHeap.poll(), maxHeap.poll()};
            Arrays.sort(three);

            minHeap.add(-three[0]);
            maxHeap.add(three[2]);

            int mid = three[1];
            if (minHeap.size() == maxHeap.size()) {
                ans.append(mid);
                minHeap.add(-mid);
            } else {
                ans.append(-minHeap.peek());
                maxHeap.add(mid);
            }
            ans.append('\n');
        }

        bw.write(ans.toString());
        bw.flush();
    }
}

File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.py
import heapq, sys
input = sys.stdin.readline

min_heap, max_heap = [], []
for i in range(int(input())):
    # max_heap에 먼저
    if i%2:
        heapq.heappush(min_heap, int(input()))
    else:
        heapq.heappush(max_heap, -int(input()))

    if min_heap and max_heap[0] < -min_heap[0]:
        heapq.heappush(min_heap, -heapq.heappop(max_heap))
        heapq.heappush(max_heap, -heapq.heappop(min_heap))

    print(-max_heap[0])